### PR TITLE
CLN: catch less in pd.io

### PR DESCRIPTION
**Pull request by jbrockmendel**
@bashtage (only tangentially related to this PR) there is an `except Exception` in pandas.io.stata that I'd like to make more specific.  Any suggestions?

### CODE REVIEW

**Comment by bashtage** [comment]
This is an intentional catch all that ensures the file gets cleanly closed.  It could legitimately encounter a wide range of errors include memory exceptions, io exceptions, windows exceptions and probably others. The only reason why Exception is caught is so it can re reraised. 

The `Exception` isn't actually needed since a bare `raise` will raise the origin exception, although if you remove it you will then have an except with no specified exception, which some don't like.

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -3271,24 +3271,29 @@ def converter(*date_cols):
...
+                        dcs, parser=date_parser, dayfirst=dayfirst
                     )
                 except Exception:
+                    # Since `date_parser` is user-provided, we can't guess
+                    #  what it might raise.
                     return generic_parser(date_parser, *date_cols)
+                else:
+                    return tools.to_datetime(parsed, errors="ignore")
+
+            else:
```
Comment (jreback) [inline pandas/io/parsers.py]: this looks like a new path, is it actually hit?

this code seems to have gotten more complex as well.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/io/parsers.py]: The to_datetime call below corresponds to the to_datetime call on 3274 in master.  This is just moving it outside of the try/except.

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -3271,24 +3271,29 @@ def converter(*date_cols):
...
-                        ),
-                        errors="ignore",
+                    parsed = parsing.try_parse_dates(
+                        dcs, parser=date_parser, dayfirst=dayfirst
                     )
                 except Exception:
+                    # Since `date_parser` is user-provided, we can't guess
+                    #  what it might raise.
                     return generic_parser(date_parser, *date_cols)
+                else:
```
Comment (jreback) [inline pandas/io/parsers.py]: do we need this or the bottom else at all?

IOW if you just remove these and have the last line of the function do what you do on 3293 right now it seems cleaner?

Comment (jbrockmendel) [reply pandas/io/parsers.py]: this isnt obvious, largely because we are dealing with a user-provided `date_parser` function.  ill take another look

**Comment by jbrockmendel** [comment]
Closing to clear the queue

**Pull request closed without merging**